src/commands/compile.rs

```diff
@@ -0,0 +1,413 @@
+//! The module that implements the `wasmtime wast` command.
+
+use crate::{init_file_per_thread_logger, CommonOptions};
+use anyhow::{anyhow, bail, Context, Result};
+use std::fs::{self, File};
+use std::io::BufWriter;
+use std::path::PathBuf;
+use structopt::{
+    clap::{AppSettings, ArgGroup},
+    StructOpt,
+};
+use target_lexicon::Triple;
+use wasmtime::{Config, Engine, Module};
+
+/// Compiles a WebAssembly module.
+#[derive(StructOpt)]
+#[structopt(
+    name = "compile",
+    version = env!("CARGO_PKG_VERSION"),
+    setting = AppSettings::ColoredHelp,
+    group = ArgGroup::with_name("x64").multiple(true),
+    group = ArgGroup::with_name("preset-x64"),
+    group = ArgGroup::with_name("aarch64").multiple(true).conflicts_with_all(&["x64", "preset-x64"]),
+    group = ArgGroup::with_name("preset-aarch64").conflicts_with_all(&["x64", "preset-x64"]),
+    after_help = "By default, no CPU flags will be enabled for the compilation.\n\
+                  \n\
+                  Use the various preset and CPU flag options for the environment being targeted.\n\
+                  \n\
+                  Usage examples:\n\
+                  \n\
+                  Compiling a WebAssembly module for the current platform:\n\
+                  \n  \
+                  wasmtime compile example.wasm
+                  \n\
+                  Specifying the output file:\n\
+                  \n  \
+                  wasmtime compile -o output.cwasm input.wasm\n\
+                  \n\
+                  Compiling for a specific platform (Linux) and CPU preset (Skylake):\n\
+                  \n  \
+                  wasmtime compile --target x86_64-unknown-linux --skylake foo.wasm\n"
+)]
+pub struct CompileCommand {
+    #[structopt(flatten)]
+    common: CommonOptions,
+
+    /// Enable support for interrupting WebAssembly code.
+    #[structopt(long)]
+    interruptable: bool,
```

Comment:
This sort of makes me think that we should lean towards `-C` flags which affect codegen. In any case though this might be good to lift into `CommonOptions` perhaps?

Reply:
Lifting it into `CommonOptions` means `wasmtime run --interuptable foo.cwasm` would be possible, which to me doesn't make nearly as much sense as the `--wasm-timeout` option specific to the run command that inherently enables interruption in a useful way.

I think `wasmtime compile --interruptable foo.wasm` and then `wasmtime run --wasm-timeout 60s foo.cwasm` would make sense to users (the latter errors if `--interruptable` wasn't given to the compile command).

That said, though, I could be convinced that `wasmtime run --interruptable foo.cwasm` is a way to run a AOT module that was compiled with interruption support without specifying a timeout, similar to `--wasm-timeout <some-reaaaaaally-long-timeout>`.